Apply tiled functionality for tiled processing of graphs

JoOkuma opened the pull request:
@yfukai, could you review this PR?

I have a function that cannot handle the whole graph, so I'm adding functionality to chunk the data.

Because we might have tile artifacts, it needs an overlay. The only way I thought of being able to do this properly is by providing two graphs to the `MapFunc`. I also considered providing the tile coordinates in case a user might want to slice the image at that position.

yfukai commented on src/tracksdata/functional/_apply.py:
```diff
@@ -0,0 +1,238 @@
+        By default "t", "z", "y", "x" are included. If some columns are not present, they will be ignored.
+    """
+
+    tile: tuple[int, ...]
```
Could be `tile_shape` for clarity?

yfukai commented on src/tracksdata/functional/_apply.py:
```diff
@@ -0,0 +1,238 @@
+    eps = 1e-8  # adding eps because np.arange is right exclusive
+    tiles_corner = list(
+        itertools.product(
+            *[np.arange(s, e + eps, t).tolist() for s, e, t in zip(start, end, tiling_scheme.tile, strict=True)]
```
Since start and end are ints, maybe we can do something like:
It may also be useful to note whether indices are inclusive or exclusive.

```suggestion
            *[np.arange(s, e+1, t).tolist() for s, e, t in zip(start, end, tiling_scheme.tile, strict=True)]
```

yfukai commented on src/tracksdata/functional/_apply.py:
```diff
@@ -0,0 +1,238 @@
+    # if agg_func is provided, we need to reduce the results
+    if tiling_scheme.attrs is None:
+        # default attrs
+        attr_keys = ["t", "z", "y", "x"]
```
Maybe we can use `DEFAULT_ATTRS` for "t".

yfukai reviewed (changes requested):
Almost all looks nice to me! I'm just afraid that the logic around `eps` might not work as expected if  `start` and `end` can accept floats (since the tile size may be arbitrary small depending on the length unit). If those are ints as noted as type annotations, maybe we can simply use `-1` or `+1` to switch exclusive / inclusive slices.

yfukai commented:
I'm sorry to be late, this is actually my first requested review

JoOkuma commented:
Hi, @yfukai, thank you for the review, and don't need to apologize.
I understand we have other things going on besides this.
I'll also finish reviewing your PR soon.

I changed `tile|overlap` to `{tile|overlap}_shape`. 

The tiling can be floats, so I changed the typing to reflect that. I also added a test to check if the scale is affecting the slicing, and I had to change the `eps` so it varies with the scale of the tile due to numerical errors.

The two `eps` are different:
- the one in the `np.arange` is because in the case when the `min` and `max` value of an attr were equal, it was returning a zero-length array.
- The slicing `eps` interacts with `spatial-graph`, which suffers from numerical error, so now it's scaled, and it works for integers and floats as covered by the new test.

Let me know if something is not clear

yfukai commented:
Hi @JoOkuma, thanks! I found out [np.nextafter](https://numpy.org/doc/2.3/reference/generated/numpy.nextafter.html) to swap the inclusive and exclusive ranges. Would it be safer to replace `eps` with this?

JoOkuma commented:
> Hi @JoOkuma, thanks! I found out [np.nextafter](https://numpy.org/doc/2.3/reference/generated/numpy.nextafter.html) to swap the inclusive and exclusive ranges. Would it be safer to replace `eps` with this?

Hi @yfukai, thanks for the input. I didn't know this function. Using `nextafter` didn't work that well with `np.arange` but I was able to get around with an `if` condition, see comment above.

For the other use case, either that or `np.spacing` leads to issues with the RTree from spatial-graph. It behaves a bit differently than we expect, so even when I tried with np.float32 precision for the tolerance, it doesn't pass the tests.

I'm considering reverting this last commit https://github.com/royerlab/tracksdata/pull/183/commits/6988d911f7635a95f3ccad1ca3152e2a9e3d1754 and leaving it as it is. What do you think?

JoOkuma commented:
Sounds good @yfukai, let me know when I can review #189
